crates/wasmtime/src/types.rs

```diff
@@ -90,14 +102,101 @@ impl ValType {
             WasmType::F32 => Self::F32,
             WasmType::F64 => Self::F64,
             WasmType::V128 => Self::V128,
-            WasmType::FuncRef => Self::FuncRef,
-            WasmType::ExternRef => Self::ExternRef,
+            WasmType::Ref(rt) => Self::Ref(RefType::from_wasm_ref_type(&rt)),
+            WasmType::Bot => Self::Bot,
         }
     }
 }
 
-// External Types
+/// A reference type holds what it refers to and whether it is nullable
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct RefType {
+    /// Indicates whether the reference is nullable.
+    pub nullable: bool,
+    /// The reference's heap type.
+    pub heap_type: HeapType,
+}
+
+impl fmt::Display for RefType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &FUNC_REF => write!(f, "funcref"),
+            &EXTERN_REF => write!(f, "externref"),
+            RefType {
+                nullable,
+                heap_type,
+            } => {
+                if *nullable {
+                    write!(f, "(ref null {})", heap_type)
+                } else {
+                    write!(f, "(ref {})", heap_type)
+                }
+            }
+        }
+    }
+}
+
+impl RefType {
+    pub(crate) fn to_wasm_ref_type(&self) -> WasmRefType {
+        WasmRefType {
+            nullable: self.nullable,
+            heap_type: HeapType::to_wasm_heap_type(&self.heap_type),
+        }
+    }
+
+    pub(crate) fn from_wasm_ref_type(rt: &WasmRefType) -> Self {
+        RefType {
+            nullable: rt.nullable,
+            heap_type: HeapType::from_wasm_heap_type(&rt.heap_type),
+        }
+    }
+}
 
+/// A list of all possible heap types in WebAssembly
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum HeapType {
+    /// A reference to a Wasm function.
+    Func,
+    /// A reference to opaque data in the Wasm instance.
+    Extern,
+    /// A typed reference to a Wasm function.
+    Index(u32),
```

Comment:
At the layer of the Wasmtime embedding here it's unfortunately not longer suitable to use a `u32` here since there's not actually any way to translate this `u32` to an actual type to inspect.

This actually raises a particularly interesting embedding question for me, though, which is how an embedder is expected to invoke an exported function from a wasm module which takes a typed function reference as an argument (or returns such a result).

I can talk more with y'all if you'd like for something like this, but the `u32` here more-or-less needs to become something like a `FuncType`, but not literally that given how lightweight `ValType` is supposed to be relative to hwo heavyweight `FuncType` is. We'll probably want some form of indexing or such but this is moving into the realm of relatively large-ish changes which y'all may want to defer to later if possible.

Reply:
I'll bump this again, this is a serious issue which must be addressed before the PR is merged. As I mentioned before though this has far-reaching implications and is unfortunately not going to be easy to handle. For example the `PartialEq`/`Eq` implementations for this type (and transitively for `ValType`) aren't valid any more to go through `u32` since two different modules could have the same `u32` pointing at different types.

Reply:
So would we need to equip the type with some sort of provenance that can inform us about the origin module?

Reply:
Yeah that's reasonable to me and I think would work out well. You may have to add a few more error-paths in places that may be infallible but that's ok, at the top-level I think everything is fallible to communicate the necessary errors here so it'd just be internal refactoring.